### PR TITLE
SAK-18587 - Synoptic tool for assignments

### PR DESCRIPTION
**Pull request by jonespm**
Old patch we had around updated for 12

### CODE REVIEW

File: assignment/assignment-tool/tool/src/webapp/vm/assignment/chef_assignments_list_assignments.vm [hunk 1]
```diff
@@ -239,7 +245,8 @@
 								#end
 							</td>						
 							<td headers="title" scope="row">
-									#if (($!allowAddAssignment || $!allowUpdateAssignment || $!service.allowGradeSubmission($assignment.getReference())) && $!view.equals('lisofass1'))
+									## These functions could be improved to improve the instructor synoptic view
+									#if (!$!isSynopticTool && ($!allowAddAssignment || $!allowUpdateAssignment || $!service.allowGradeSubmission($assignment.getReference())) && $!view.equals('lisofass1'))
```
Comment (buckett) [inline assignment/assignment-tool/tool/src/webapp/vm/assignment/chef_assignments_list_assignments.vm]: Is this change needed as you're setting `$allowAddAssignment` to `false` at the top of the template.

File: assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 2]
```diff
@@ -14530,9 +14538,22 @@ else if (deleted.equalsIgnoreCase(Boolean.TRUE.toString()) && (a.getContent().ge
 			}
 			else
 			{
-				// read all Assignments
-				returnResources = assignmentService.getListAssignmentsForContext((String) state
-						.getAttribute(STATE_CONTEXT_STRING));
+				if (isSynopticTool() & isOnWorkspaceTab()) {
```
Comment (buckett) [inline assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Indentation looks wrong here.

File: assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 3]
```diff
@@ -14530,9 +14538,22 @@ else if (deleted.equalsIgnoreCase(Boolean.TRUE.toString()) && (a.getContent().ge
...
+					//Get all sites for this user
+					String userId = sessionManager.getCurrentSessionUserId();
+	                List<Site> sites = siteService.getSites(SelectionType.ACCESS, null, null, null,null, null);
+	                for (Site site : sites) {
+	                	Member sm = site.getMember(userId);
+	                	if (sm != null) {
+	                		returnResources.addAll(assignmentService.getListAssignmentsForContext(site.getId()));
```
Comment (buckett) [inline assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: While this works might be worth pushing this up into the API so that you can pass in a list of `siteIds` to get assignments for, rather than going back to the DB for each one.

Also without paging could this end up causing problems where archiving of old course sites doesn't happen quickly and a tutor can end up with 100s of assignments being retrieved.

File: assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java [hunk 4]
```diff
@@ -17743,6 +17764,41 @@ protected void letterGradeOptionsIntoContext(Context context) {
 		context.put("letterGradeOptions", StringUtils.split(lOptions, ","));
 	}
 
+	/**
```
Comment (buckett) [inline assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Indentation looks off.

File: assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java (hunk 3 above)
Comment (jonespm) [inline assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Yeah, I did this back in something like 2011 and for sure didn't think of the performance implications. I'll see what can be done about this.

Comment (buckett) [reply assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: Seen any production issues from this at all?

Comment (jonespm) [reply assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java]: No we never did, but only ran it on a few clients that probably didn't have a lot of sites. It looks like there was a method added here by SAK-23358 (getMyAssignmentsForAllSite) that does some of this already and a few other tools do this too. I feel like any large site enrollment could cause an issue. Maybe it should just be more limited.

**Comment by jonespm** [comment]
I'm just going to close this PR for now. It needs quite a few updates for the new assignments and it needs this performance fix Buckett is talking about. I'm not going to be working on this soon.
